core/src/main/java/org/jruby/RubyEnumerator.java

```diff
@@ -519,91 +524,26 @@ public IRubyObject with_index19(ThreadContext context, IRubyObject arg, final Bl
-        return nexter.next();
-    }
-
-    @JRubyMethod
-    public synchronized IRubyObject rewind(ThreadContext context) {
-        if (object.respondsTo("rewind")) object.callMethod(context, "rewind");
-
-        if (nexter != null) {
-            nexter.shutdown();
-            nexter = null;
-        }
-
-        return this;
-    }
-
-    @JRubyMethod
-    public synchronized IRubyObject peek(ThreadContext context) {
-        final Nexter nexter = ensureNexter(context.runtime);
-
-        return nexter.peek();
-    }
-
-    @JRubyMethod(name = "peek_values")
-    public IRubyObject peekValues(ThreadContext context) {
-        return RubyArray.newArray(context.runtime, peek(context));
-    }
-
-    @JRubyMethod(name = "next_values")
-    public IRubyObject nextValues(ThreadContext context) {
-        return RubyArray.newArray(context.runtime, next(context));
-    }
+    // java.util.Iterator :
 
-    @JRubyMethod
-    public synchronized IRubyObject feed(ThreadContext context, IRubyObject val) {
-        final Nexter nexter = ensureNexter(context.runtime);
-        if (!feedValue.isNil()) {
-            throw context.runtime.newTypeError("feed value already set");
-        }
-        feedValue = val;
-        nexter.setFeedValue(val);
-        return context.nil;
-    }
+    @Override
+    public synchronized boolean hasNext() {
+        Ruby runtime = getRuntime();
 
-    private Nexter ensureNexter(final Ruby runtime) {
-        Nexter nexter = this.nexter;
-        if (nexter != null) return nexter;
+        IRubyObject generator = getInstanceVariable("@__generator__");
 
-        if (Options.ENUMERATOR_LIGHTWEIGHT.load()) {
-            if (object instanceof RubyArray && method.equals("each") && methodArgs.length == 0) {
-                return this.nexter = new ArrayNexter(runtime, object, method, methodArgs);
-            }
+        if (generator == null || generator == runtime.getNil()) {
+            return false;
         }
-        return this.nexter = new ThreadedNexter(runtime, object, method, methodArgs);
-    }
 
-    @Override
-    protected void finalize() throws Throwable {
-        try {
-            Nexter nexter = this.nexter;
-            if (nexter != null) {
-                nexter.shutdown();
-                nexter = null;
-            }
-        } finally {
-            super.finalize();
-        }
-    }
-
-    // java.util.Iterator :
-
-    @Override
-    public synchronized boolean hasNext() {
-        return ensureNexter(getRuntime()).hasNext();
+        ThreadContext context = runtime.getCurrentContext();
+        return sites(context).next_p.call(context, this, generator).isTrue();
     }
 
     @Override
     public Object next() {
-        return next( getRuntime().getCurrentContext() ).toJava( java.lang.Object.class );
+        ThreadContext context = getRuntime().getCurrentContext();
+        return sites(context).next_p.call(context, this, this).toJava( java.lang.Object.class );
```

Comment:
you might have wanted `sites(context).next.call(...)` here to invoke `next` (not `next?`)

Reply:
You are correct sir, good catch. Must not be tested.

Reply:
Looks like this is tested and failed. I'll finish up the PR and make sure it's all green.